### ISSUE TITLE
Create proposal for standardized validation

### ISSUE THREAD

**Issue by rbennettcw** [hicommonwealth/commonwealth#5212]
Currently, we don't apply a consistent pattern for validating data.

Things we need to validate:
- Requests on the backend (query params, body, etc.)
- Forms on the frontend

Recently, some of the newer routes on the backend (e.g. API Gating) use the Zod library for request validation.
The frontend has also recently used the Zod library for form validation, but it has been problematic for a specific usecase.

Ideally, we should use the same JS/TS validation library for both the frontend and backend.

# Acceptance Criteria
- Choose which library we will use for validation
- Create a wiki document outlining the decision w/ code examples
- Create new ticket(s) to update any parts of the codebase affected by this change

**Comment by rbennettcw** [comment]
Since we need to apply validation on both the frontend and backend, perhaps we can create a standard for validation across the codebase?

@mzparacha @masvelio @jnaviask

**Comment by mzparacha** [comment]
I agree, but I think initially we would have some changes as we develop the modules. Personally, I would like to wait for some initial work to go in, so we can better learn what would work, but that's just me <- just sharing my thoughts, could be wrong.

**Comment by kurtisassad** [comment]
For backend we use express-validator as a library for some routes. I am open to refactoring this all to another error handling library if we decide on one. I like express-validator because it separates the error handling from the route itself, so it makes it easier to reason with/doesn't make the code as messy.

I am not sure if we can have common validation between the back and front end, because the front end validation needs to interact with react's state and treat each state variable differently. In which case having common validation would be tricky to abstract out.

**Comment by kurtisassad** [comment]
Ok looking more into things I think this comment is correct and doable:
> would be worth to abstract out the common logic

Reading more into the react world, people use two libraries in conjunction:
* [zod](https://www.npmjs.com/package/zod) (Used for schema validation)
* [react-hook-form](https://www.npmjs.com/package/react-hook-form) (Used for zod integration with react state)

With these two libraries we would have as follows:
Common-commons: zod schemas
Front-end: imports zod schema, passes them into useForm from react-hook-form.
Back-end: imports zod schema on the request (maybe needs to augment the schema for other request parameters)

So basically for each post request (where we take data from a form and pass it into the backend) we can abstract out the error validation and message handling with zod. If we were to go with this path it would be worthwile to remove the express-validator library, as we would be able to unify validation.

This is just a rough proof of concept as I have not used these libraries or have tried this before. Bringing @Rotorsoft into the discussion as he has worked with zod before (or at least was discussing it during our interview).

**Comment by Rotorsoft** [comment]
In terms of schema validation, zod is probably the winner today. Not the fastest but fast enough, great TS integration, small footprint, and healthy ecosystem (hook-forms resolvers, OpenAPI spec generators, Express routes, JSON schema, etc). We should probably pick a small piece of new end-to-end features and use it as a POC. The front/back end integration is tricky as noted above, not sure if you have already discussed pros/cons of different strategies (proxy generators vs shared contracts, etc).
